api/cache/objectslist.go

```diff
@@ -0,0 +1,113 @@
+package cache
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/bluele/gcache"
+	cid "github.com/nspcc-dev/neofs-api-go/pkg/container/id"
+	"github.com/nspcc-dev/neofs-api-go/pkg/object"
+)
+
+/*
+	This is an implementation of a cache which keeps unsorted lists of objects' IDs (all versions)
+	for a specified bucket and a prefix.
+
+	The cache contains gcache whose entries have a key: ObjectsListKey struct and a value: list of ids.
+	After putting a record it lives for a while (default value is 60 seconds).
+
+	When we receive a request from the user we try to find the suitable and non-expired cache entry, go through the list
+	and get ObjectInfos from common object cache or with a request to NeoFS.
+
+	When we put an object into a container we invalidate entries with prefixes that are prefixes of the object's name.
+*/
+
+type (
+	// ObjectsListCache provides interface for cache of ListObjectsV2 in a layer struct.
+	ObjectsListCache interface {
+		Get(key ObjectsListKey) []*object.ID
+		Put(key ObjectsListKey, oids []*object.ID) error
+		CleanCacheEntriesChangedWithPutObject(objectName string, cid *cid.ID)
+	}
+)
+
+const (
+	// DefaultObjectsListCacheLifetime is a default lifetime of entries in cache of ListObjects.
+	DefaultObjectsListCacheLifetime = time.Second * 60
+	// DefaultObjectsListCacheSize is a default size of cache of ListObjects.
+	DefaultObjectsListCacheSize = 1e5
+)
+
+type (
+	// ListObjectsCache contains cache for ListObjects and ListObjectVersions.
+	ListObjectsCache struct {
+		lifetime time.Duration
```

Comment:
I suppose we can remove this field if we will create gcache like:
```
gcache.New(cacheSize).LRU().Expiration(lifetime)
```

Reply:
I'm not sure it's necessary. I don't think lifetime takes so much memery but it leaves the opportunity to change lifetime if we need.
Or maybe we should remove this from other caches: buckets, system etc.?

Reply:
It's not about memory. I just think that if we use `gcache` and it has functionality that we need we should use this feature.